### ISSUE TITLE
Json error in email attachment logic apps

### ISSUE THREAD

**Issue by joanslow** [gsureshqa/LogicApp#3]
The notes app is installed & running successfully.
However both email attachment apps fail with the error
"InvalidTemplate. Unable to process template language expressions in action 'CreateAzureBlob' inputs at line '1' and column '2769': The template expects its parameter to be a string or an XML. The provided value is of type 'Object'. Please see https://aka.ms/logicexpressions#json for usage details."

Please could you advise how to get them working? Many thanks.

![attachmentlogicapperror](https://user-images.githubusercontent.com/9998155/39933260-60082846-553a-11e8-8f6c-d7b818ebed8c.png)

**Comment by anilvem1** [comment]
Please replace the below line via Logic App -> Code View
OLD: "name": "@{concat(concat(items('For_each')?['activitymimeattachmentid'], '_'), json(items('For_each')?['attachmentid'])['filename'])}"

NEW: "name": "@{concat(items('For_each')?['activitymimeattachmentid'], '_', items('For_each')?['attachmentid']?)['filename']}"

We will soon publish the updated Logic App.

**Comment by joanslow** [comment]
Anil,
Thanks for the update.
In the app code I found the suggested line, but with For_each_2, not
For_each. However, adding your suggested line prevents the code from being
saved. Moving the final bracket as below, allows for save:

"name": "@{concat(items('For_each_2')?['activitymimeattachmentid'], '_',
items('For_each_2')?['attachmentid']?['filename'])}"

However, it is still sometimes failing with the same error. The only other
line referencing json is

"body":
"@base64ToBinary(json(items('For_each_2')?['attachmentid'])['body'])",

Many thanks,
Jo

**Comment by anilvem1** [comment]
Yes please. Replace the other line also. Just remove json ahead.
replace the line 
OLD: json(items('For_each')?['attachmentid'])['filename']
NEW: items('For_each')?['attachmentid']['filename']

**Comment by rerklin** [comment]
Hi Anil,

I was getting the same issues as Jo but your last post has changes that don't seem to be working. I updated to:
OLD: "@base64ToBinary(json(items('For_each')?['attachmentid'])['body'])",
NEW: "@base64ToBinary(items('For_each')?['attachmentid']['filename'])",

However, I am still getting the following error when running it:
![image](https://user-images.githubusercontent.com/28170492/42600840-6719c292-8529-11e8-8e52-76ccc718c1e7.png)

Thanks!
Ryan

**Comment by anilvem1** [comment]
Please use the below. This has the updated code using CRM Web API.

https://github.com/anilvem1/LAMoveCRMAttachmentsToBlob-API

**Comment by rerklin** [comment]
Ok, I'll have to go through this new setup tomorrow as I have to read more about setting up this CRM Web API.

Thanks!

**Comment by rerklin** [comment]
Is there any way to just update what I already have or will that not work with Dynamics 365 v9? I just want to move over all existing email attachments after adding "Attachment Management". This looks like a lot of work for something I will just run once. 

Thanks for your work!

**Comment by anilvem1** [comment]
NEW: "@base64ToBinary(items('For_each')?['attachmentid']['body'])",

Replace filename with body

**Comment by rerklin** [comment]
Thanks Anil! That worked perfectly. 

One last question, with the $top parameter. Is there a limit to the number you'd recommend putting there? I have about 100,000 email attachments to move. They're not large in size, but there are a lot. Most likely email signature files. 

Thanks Again!

Ryan

**Comment by anilvem1** [comment]
You can remove $top statement.
Go to LA designer -> GetAttachments action -> ... -> settings -> enable paging and give count as you desire (preferred 5k or 10k) and click Save
Save LA and run.

**Comment by rerklin** [comment]
Thanks Anil, I'll give that a try. For now though, I just shortened the recurrence and it's going along. 

I did run into another issue though, I am getting an error "A record with matching key values already exists.".  Any ideas?
![image](https://user-images.githubusercontent.com/28170492/42654548-98b2a566-85de-11e8-9567-75752f07273a.png)
